push-agents/apns-agent/src/main/java/org/aerogear/push/apns/ios/SimpleApnsClientCache.java

```diff
@@ -0,0 +1,153 @@
+import net.jodah.expiringmap.ExpirationPolicy;
+import net.jodah.expiringmap.ExpiringMap;
+import org.jboss.aerogear.unifiedpush.api.iOSVariant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+public final class SimpleApnsClientCache {
+
+    private static final SimpleApnsClientCache INSTANCE = new SimpleApnsClientCache();
+
+    private SimpleApnsClientCache() {
+        // no-op
+    }
+
+    public static SimpleApnsClientCache getInstance(){
+        return INSTANCE;
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(SimpleApnsClientCache.class);
+
+    final ConcurrentMap<String, ApnsClient> apnsClientExpiringMap;
+    {
+        apnsClientExpiringMap = ExpiringMap.builder()
+
+                // TODO: would be nice if it could be configured via System property:
+                .expiration(12, TimeUnit.HOURS)
+                .expirationPolicy(ExpirationPolicy.ACCESSED)
+                .asyncExpirationListener((ExpirationListener<String, ApnsClient>) (variantID, apnsClient) -> {
+
+                    if (apnsClient.isConnected()) {
+                        logger.info("APNs connection for iOS Variant ({}) was inactive last 12 hours, disconnecting...", variantID);
+
+                        final Future<Void> disconnectFuture = apnsClient.disconnect();
+
+                        disconnectFuture.addListener(future -> {
+
+                            if (future.isSuccess()) {
+                                logger.debug("Disconnected from APNS due to inactive connection for iOS Variant ({})", variantID);
+                            } else {
+                                final Throwable t = future.cause();
+                                logger.warn(t.getMessage(), t);
+                            }
+                        });
+                    }
+                }).build();
+    }
+
+    public ApnsClient getApnsClientForVariant(final iOSVariant iOSVariant, final ServiceConstructor<ApnsClient> constructor) {
+        final String connectionKey = extractConnectionKey(iOSVariant);
+        ApnsClient client = apnsClientExpiringMap.get(connectionKey);
+
+        if (client == null) {
+            logger.debug("no cached connection for {}, establishing it", connectionKey);
+            synchronized (apnsClientExpiringMap) {
+                client = constructor.construct();
+
+                if (client.isConnected()) {
+                    putApnsClientForVariantID(connectionKey, client);
+                }
+
+                return client; // return the newly connected client
+            }
+        } else {
+            logger.debug("reusing cached connection for {}", connectionKey);
+            return client; // we had it already
+        }
+    }
+
+//    public void disconnectOnChange(@Observes final iOSVariantUpdateEvent iOSVariantUpdateEvent) {
```

Comment:
Can we remove this?

Reply:
nope - this, again, is POC code - if I remove it, I might forget that this is needed in the long run